[hl] use parallel run for opt / write_c

yuxiaomao opened the pull request:
Trying to use Ocaml 5 parallel programming feature on hl.

Isolate hlopt phase and run with `domainslib`. Use `-D domains=N` to set number of domains (default to 4 just to demonstrate that it works).

The hlopt dump result `hlopt.txt` is not longer guaranteed to be the same in different runs (functions body remains the same but their order can change).

My best result for now is around `1/2` hlopt time gain with `16` threads (my CPU have 16 logical cores): 3.6s (1 domains) -> 3.3s (4 domains) ->  2.6s (8 domains) -> 1.5s (16 domains), in a total compilation time ~50s. I think I'm bound by shared data / IO.
When using `-D dump` I have a better gain on hlopt: 15s -> 4s (16 domains), in a total time 66s -> 56s.

Related work on hlopt cache (but it is not supported by my current implementation, and is not thread safe)
- https://github.com/HaxeFoundation/haxe/pull/11797

Note:
- `DynArray.get` is probably not thread safe as described in the doc `Alert unsynchronized_access`, perhaps I need a mutex; but it didn't crash yet.
- Thread pool can be hold by global context (`com`) if more modules might need it.
- Choice of `~chunk_size:16` : it just works better than `default/8/32` on my PC, but [some tutorial](https://github.com/ocaml-multicore/parallel-programming-in-multicore-ocaml) says this too 
> ... On the other hand, if the amount of time taken is different for every iteration, the chunks should be smaller. If the total number of iterations is a sizeable number, a chunk_size like 32 or 16 is safe to use ...

yuxiaomao committed:
[hl] use parallel run for opt

yuxiaomao committed:
Use Ocaml 5 on windows

yuxiaomao committed:
[hlc] parallel write .c files

yuxiaomao committed:
Use Ocaml 5 only on Linux

Simn commented:
> * Thread pool can be hold by global context (`com`) if more modules might need it.

This sounds like a good idea to me, so if that's best practice then let's do it. It would also be good to have a convenience function like `run_parallel com length f` which wraps the `Domainslib.Task.run` thing. If we ever need a different value for `start` we can refine that later.

Edit: A good place for that should be a new src/context/parallel.ml.

yuxiaomao committed:
Move mkdir_recursive to sequential domain

yuxiaomao committed:
Merge branch 'development' into dev-hlopt-parallel

yuxiaomao committed:
Move parallel to src/context, remove chunk_size, adjust mkdir/buffer

yuxiaomao commented:
I've added the parallel support in `hl2c.write_c`, and moved the parallel part into `src/context/parallel.ml`.
I've also removed `chunk_size` because with the default domain count, the default `chunk_size` works good enough.

However adding the parallel in global context increase macro time x2, I'm still investigating.

### hl2c

**Drawback**: The hash order in `hl/hashes.c` is no longer guaranteed to be the same in different runs (the calls remains the same but their order can change, which may result in same string have different hash in different runs in the case of hash collision).

**My best result for now**: around `1/3` hlwrite time gain with `24` threads (default domain count) fresh-gen/re-gen: 10s/8s (1 domain) -> 7s/5s (24 domains), in a total compilation time ~60s.

yuxiaomao committed:
Create pool on-demande is quicker

Simn commented:
Note for future endeavors: our unification isn't thread-safe due to the global stacks: 

```ocaml
let unify_stack = new_rec_stack()
let variance_stack = new_rec_stack()
let abstract_cast_stack = new_rec_stack()
let unify_new_monos = new_rec_stack()
```

I just ran into this from genjvm attempting to unify something only to choke on a `List.tl abstract_cast_stack`. This could affect tanon_identification and thus hxb as well @kLabz.

yuxiaomao commented:
I tried implement parallel on genhl, but there are too many shared states/index tables/cache etc that I didn't fully resolved yet.
I'm wondering if it useful at all: too many mutex + can no longer guarantee the relatively same result on different runs (or have a big overhead for re-ordering and patching things).

yuxiaomao committed:
Merge branch 'development' into dev-hlopt-parallel

Simn commented:
Yeah, HL's context doesn't look very suitable for parallelism with all these lookups and whatnot...

yuxiaomao committed:
Fix syntax and cleanup some comment

yuxiaomao committed:
Reduce change and use Mutex.protect

Simn commented on src/generators/genhl.ml:
```diff
@@ -86,21 +86,21 @@ type constval =
 	cnatives : (string * int, (string index * string index * ttype * functable index)) lookup;
 	cfids : (string * path, unit) lookup;
-	cfunctions : fundecl DynArray.t;
+	cfunctions : (fundecl * bool) DynArray.t;
```
I don't reaaaaally care but these non-descriptive bools can be quite difficult to make sense of. I generally try to go for a `{ what_is_this: bool }` style, even if it's just a single field. That also makes it easier to add more information later, should the need arise.

yuxiaomao replied:
I would like to (here `bool = need_opt`), but I don't know how to add a description/label to the field (I didn't find similar description in haxe code)? Maybe just add a comment next to it?

Simn replied:
We don't need actual descriptions, the name is usually enough.

Simn commented on src/generators/hl2c.ml:
```diff
@@ -56,29 +56,34 @@ and function_entry = {
-	mutable file_prefix : string;
 	mutable fun_index : int;
-	mutable type_module : (ttype, code_module) PMap.t;
-	gcon : Gctx.t;
```
Hmm, this shouldn't be necessary because gctx is pretty const.

Simn replied:
Ah never mind, I thought you were passing this context around further below, but that's actually a different context also named gctx.

yuxiaomao committed:
Move need_opt into fundecl

Simn committed:
Merge branch 'development' into dev-hlopt-parallel

# Conflicts:
#	src/generators/genhl.ml

yuxiaomao commented:
I would like to merge this if you're ok with it too.
